py/foundation/geometry.py

```diff
@@ -244,3 +285,96 @@ def distance(b1: 'BBox', b2: 'BBox') -> float:
     inner_width = max(0, ix.length - b1.ix.length - b2.ix.length)
     inner_height = max(0, iy.length - b1.iy.length - b2.iy.length)
     return sqrt(inner_width**2 + inner_height**2)
+
+
+@dataclass(frozen=True)
+class Separator:
+  start_point: Point
+  end_point: Point
+
+  def min_distance_from_point(self, p: 'Point'):
+    '''
+    Minimum distance between a line from point a to point b and point e
+
+    >>> Separator(Point(0,0),Point(10,0)).min_distance_from_point(Point(-3,4))
+    5.0
+    >>> Separator(Point(0,0),Point(10,0)).min_distance_from_point(Point(16,8))
+    10.0
+    >>> Separator(Point(0,0),Point(10,0)).min_distance_from_point(Point(7,6))
+    6.0
+    '''
+    a, b = self.start_point, self.end_point
+    ab = b - a
+    bp = p - b
+    ap = p - a
+    ab_bp = ab @ bp
+    ab_ap = ab @ ap
+
+    if ab_bp > 0:
+      return (bp@bp) ** (1/2)
+    elif ab_ap < 0:
+      return (ap@ap) ** (1/2)
+    else:
+      return abs((ab.x * ap.y) - (ab.y * ap.x)) / (ab@ab) ** (1/2)
+
+  def intersects(self, other: 'Separator', max_distance=0):
+    distance = min(
+      self._min_distance_line_point(self, other.start_point),
+      self._min_distance_line_point(self, other.end_point),
+      self._min_distance_line_point(other, self.start_point),
+      self._min_distance_line_point(other, self.end_point),
+    )
+    return distance <= max_distance
+
+  def is_vertical(self, max_d=2):
```

Comment:
Whoa, what is 2?

Reply:
Haha good catch - this is just an arbitrary hyper-parameter that app eng has been using. How would you suggest doing this kinda thing?

Reply:
It seems like an artificial problem -- is there a single real-world usage of this type where the segment is not either vertical or horizontal?

Reply:
> > > Code-wise this LGTM. But what exactly is a separator?
> > > Is a separator something in a document? In that case it should be an `Entity` and have a `BBox`.
> > > Is a separator a convenience type for telling utility code how to chop up a document into smaller documents? In that case, why the genericity? In this context separators will always be horizontal/vertical, no?
> > 
> > 
> > Separator is a generic geometrical type like Point corresponding with concept of LineSegment (could be called that, but just worried about confusion with Line as in line of text). A Separator _can_ be an entity; and it _can_ be useful for splitting; but it's also a very useful geometrical concept.
> 
> Can we just call it `LineSegment` then, since that's what this actually is?

I have no strong feeling one way or the other. I do worry about confusion, given that we commonly speak of a "line in a document" as referring to something else. Obviously "Separator" has essentially no associations,

Reply:
- I have a scanned in document with some skew that isn't perfectly resolved. So my 'vertical lines' aren't actually vertical relative to the document's coordinate system and aren't actually parallel to each other. 
- If OCR finds a line that looks diagonal, do we just say "either figure out if it's horizontal or vertical, or else throw it away"?
- If a vertical line is a thing and a horizontal line is a thing, why would we possibly not want to say "both are lines" to make code neater?